geode-core/src/test/java/org/apache/geode/cache/client/internal/ConnectionFactoryImplTest.java

```diff
@@ -100,4 +107,32 @@ public void authenticateIfRequired() {
     verify(pool).executeOn(any(Connection.class), any(Op.class));
     verify(server).setUserId(123L);
   }
+
+  @Rule
+  public ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  @Test
+  public void getSetServerIdShouldBeSynchronizedToAvoidExtraLoginRequest() throws Exception {
+    when(pool.isUsedByGateway()).thenReturn(false);
+    when(pool.getMultiuserAuthentication()).thenReturn(false);
+    when(pool.executeOn(any(Connection.class), any())).thenReturn(123L);
+    when(server.getRequiresCredentials()).thenReturn(true);
+    when(connection.getServer()).thenReturn(server);
+
+    doCallRealMethod().when(server).setUserId(anyLong());
+    doCallRealMethod().when(server).getUserId();
+    // set up the initial userId to be -1
+    server.setUserId(-1);
+
+    CompletableFuture<Void> future1 =
+        executor.runAsync(() -> factory.authenticateIfRequired(connection));
+    CompletableFuture<Void> future2 =
+        executor.runAsync(() -> factory.authenticateIfRequired(connection));
+    future1.get();
+    future2.get();
+
+    // setUserId should only be called once by this two threads. The other time
+    // is called by the test setup.
+    verify(server, times(2)).setUserId(anyLong());
```

Comment:
I don't know if it's actually better, since clearing invocations on mocks is not a best practice, but to make this verification more intuitive, you can call `Mockito.clearInvocations(server);` on line 126 and then have this verification just be: `verify(server).setUserId(anyLong());` and remove the comment explaining why it's 2.